### DISCUSSION TITLE
OLED, how to display selected bank and what knob is being turned.

### DISCUSSION THREAD

**Discussion by laktosterror** [tttapa/Control-Surface#557] [Q&A]
Greetings all,

So i have planned a midi controller with six rotary encoders (buttons builtin), a single cherry mx switch for changing banks and a ssd1306 spi oled.

Since the rotary encoders are endless i was hoping to use the screen to show "where" the encoders are and the selected bank. Im unsure how to do this, this is where im asking for your help.

As i understand the code wont need to use midi input, but rather some kind of call back from variables saved by the actual mcu/controller, but im not a seasoned programmer so im not sure.

This is my current sketch (pseudocode). This example currently uses the IncrementSelectorLEDs to display the selected bank.
I have set all pins to 0 as i have not built the controller yet.

```cpp
#include <Control_Surface.h>

USBMIDI_Interface midi; // MIDI Interface to use
 
Bank<5> bank(6);

IncrementSelectorLEDs<5> bankSelector {
  bank,
  {0},         // button pins
  {0, 0, 0, 0, 0}, // LED pins
};
 
using namespace MIDI_CC;
Bankable::CCButtonLatched<5> lat[] {
  {{bank, BankType::CHANGE_ADDRESS}, 0, {Sound_Controller_1, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, 0, {Sound_Controller_2, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, 0, {Sound_Controller_3, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, 0, {Sound_Controller_4, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, 0, {Sound_Controller_5, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, 0, {Sound_Controller_6, CHANNEL_1}},
};

Bankable::CCAbsoluteEncoder enc[] = {
  {{bank, BankType::CHANGE_ADDRESS}, {0, 0}, {Sound_Controller_7, CHANNEL_1},1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, {0, 0}, {Sound_Controller_8, CHANNEL_1},1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, {0, 0}, {Sound_Controller_9, CHANNEL_1},1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, {0, 0}, {Sound_Controller_10, CHANNEL_1},1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, {0, 0}, {General_Purpose_Controller_5, CHANNEL_1},1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, {0, 0}, {General_Purpose_Controller_6, CHANNEL_1},1, 4},
};

void setup() {
  Control_Surface.begin();
}
 
void loop() {
  Control_Surface.loop();
}
```

Any tips/pointer/help is greatly appreciated.
Take care now

**Comment by tttapa** [comment]
You could try something like this: 
```cpp
#include <Control_Surface.h>
// Include the display interface you'd like to use
#include <Display/DisplayInterfaces/DisplayInterfaceSSD1306.hpp>

// ----------------------------- Display setup ------------------------------ //

constexpr uint8_t SCREEN_WIDTH = 128;
constexpr uint8_t SCREEN_HEIGHT = 64;
constexpr int8_t OLED_DC = 8;    // Data/Command pin of the display
constexpr int8_t OLED_reset = 6; // Use the external RC circuit for reset
constexpr int8_t OLED_CS = 10;   // Chip Select pin of the display
constexpr uint32_t SPI_Frequency = 32'000'000;

// Instantiate the displays
Adafruit_SSD1306 ssd1306Display {
    SCREEN_WIDTH, SCREEN_HEIGHT, &SPI, OLED_DC, -1, OLED_CS, SPI_Frequency,
};

// --------------------------- Display interface ---------------------------- //

// Implement the display interface, specifically, the begin and drawBackground
// methods.
struct MySSD1306_DisplayInterface : SSD1306_DisplayInterface {
    MySSD1306_DisplayInterface(Adafruit_SSD1306 &display)
        : SSD1306_DisplayInterface(display) {}
    void begin() override {
        if (!disp.begin()) // Initialize the Adafruit_SSD1306 display
            FATAL_ERROR(F("SSD1306 allocation failed."), 0x1306);
        // If you override this method, remember to call the super class method
        SSD1306_DisplayInterface::begin();
    }
    void drawBackground() override {} // nothing yet
} display = ssd1306Display;

// ---------------------------- Display elements ---------------------------- //

template <uint8_t NumBanks>
struct EncoderDisplay : DisplayElement {
    EncoderDisplay(DisplayInterface &display,
                   const Bankable::CCAbsoluteEncoder<NumBanks> &enc,
                   PixelLocation loc, int16_t w, int16_t h, uint16_t color)
        : DisplayElement(display), enc(enc), loc(loc), w(w), h(h) {}

    // Should the display be updated?
    bool getDirty() const override { return enc.getValue() != value; }

    // Draw to the display
    void draw() override {
        value = enc.getValue();
        int16_t q = enc.getMaxValue() + 1;
        int16_t v = (value * uint32_t(h - 4) + q / 2) / q;
        display.drawFastVLine(loc.x, loc.y + 1, h - 2, WHITE);
        display.drawFastVLine(loc.x + w - 1, loc.y + 1, h - 2, WHITE);
        display.drawFastHLine(loc.x + 1, loc.y, w - 2, WHITE);
        display.drawFastHLine(loc.x + 1, loc.y + h - 1, w - 2, WHITE);
        display.fillRect(loc.x + 2, loc.y + h - 2 - v, w - 4, v, WHITE);
    }

  private:
    const Bankable::CCAbsoluteEncoder<NumBanks> &enc;
    PixelLocation loc;
    int16_t w, h;
    uint16_t color;
    int16_t value = 0xFFFF;
};

struct BankSettingDisplay : DisplayElement {
    BankSettingDisplay(DisplayInterface &display, const OutputBank &bank,
                       PixelLocation loc, uint8_t size, uint16_t color)
        : DisplayElement(display), bank(bank), x(loc.x), y(loc.y), size(size),
          color(color) {}

    bool getDirty() const override { return bank.getSelection() != selection; }

    void draw() override {
        selection = bank.getSelection();
        display.setTextColor(color);
        display.setTextSize(size);
        display.setCursor(x, y);
        display.print(selection + 1);
    }

  private:
    const OutputBank &bank;
    int16_t x, y;
    uint8_t size;
    uint16_t color;
    setting_t selection = NO_SETTING;
};

// ------------------------------- MIDI stuff ------------------------------- //

USBDebugMIDI_Interface midi;

Bank<5> bank(6);
IncrementSelector<5> bankSelector {bank, 4};

using namespace MIDI_CC;
Bankable::CCButtonLatched<5> lat[] {
    {bank, 5, Sound_Controller_1},
};

Bankable::CCAbsoluteEncoder<5> enc[] {
    {bank, {0, 1}, Sound_Controller_7, 1, 4},
    {bank, {2, 3}, Sound_Controller_8, 1, 4},
    // etc.
};

// Display elements
EncoderDisplay<5> encdisp[] {
    {display, enc[0], {0 + 2, 0}, 8, 64, WHITE},
    {display, enc[1], {12 + 2, 0}, 8, 64, WHITE},
    {display, enc[0], {24 + 2, 0}, 8, 64, WHITE}, // change encoder indices here
    {display, enc[0], {36 + 2, 0}, 8, 64, WHITE},
    {display, enc[0], {48 + 2, 0}, 8, 64, WHITE},
    {display, enc[0], {60 + 2, 0}, 8, 64, WHITE},
};

BankSettingDisplay bankdisp {
    display, bank, {96, 24}, 3, WHITE,
};

void setup() {
    if (OLED_reset >= 0) {
        SPI.begin();
        digitalWrite(OLED_reset, LOW);
        pinMode(OLED_reset, OUTPUT);
        delay(10);
        pinMode(OLED_reset, INPUT_PULLUP);
        delay(10);
    }
    Control_Surface.begin();
}

void loop() { Control_Surface.loop(); }
```
The result:
![IMG_20210823_160014__01](https://user-images.githubusercontent.com/21296729/130463256-6fe2dcda-bb70-4cb8-960e-87512f1e603f.jpg)

**Comment by laktosterror** [comment]
Thanks for your reply Pieter.

This is perfect!
I actually received my pcb's today:
oled and led footprints uses the same teensy pins, hence i can choose if i want to use either (needs some soldering on the backside).

Me and my band mates will use these to control sends for individual channel volumes when we're recording with headphones.
Don't mind the "branding", its an in-joke..
![image](https://user-images.githubusercontent.com/55037659/130530341-b4828ab6-c694-4491-a4e0-eda8516d8fdd.png)


With encoders, cherry switch and i2c oled (had laying around).
![image](https://user-images.githubusercontent.com/55037659/130530556-c1ba8892-74a8-486e-999a-d3ecd584fae4.png)

**Reply by laktosterror** [reply]
Im trying to figure out the code that you so kindly provided. Im not doing too well... Could i bother you for two things:
1. How would i put the bank element i the middle between the third and fourth bar: " lll 1 lll " ?

2. if you could explain the "syntax" so to speak, and maybe i can figure out the rest.
```cpp
{display,       enc[0],             {0 + 2, 0},                      8, 64,        WHITE},  
    ^variable      ^ variable+[?]       ^coordinates on display?      ^?              ^color
```

```cpp
// Display elements
EncoderDisplay<5> encdisp[] {
    {display, enc[0], {0 + 2, 0}, 8, 64, WHITE},  
    {display, enc[1], {12 + 2, 0}, 8, 64, WHITE},  <-- im usure why there is a 1 in enc[1] and not a 0.
    {display, enc[0], {24 + 2, 0}, 8, 64, WHITE}, // change encoder indices here
    {display, enc[0], {36 + 2, 0}, 8, 64, WHITE},
    {display, enc[0], {48 + 2, 0}, 8, 64, WHITE},
    {display, enc[0], {60 + 2, 0}, 8, 64, WHITE},
```

Thanks again Pieter!

**Reply by tttapa** [reply]
The constructor arguments are declared higher up in the sketch: 
```cpp
    EncoderDisplay(DisplayInterface &display,
                   const Bankable::CCAbsoluteEncoder<NumBanks> &enc,
                   PixelLocation loc, int16_t w, int16_t h, uint16_t color)
```
`display` is the OLED display to draw to (you can have multiple didplays), `enc` is the encoder you want to show the value of, `loc` is the (x,y) position of the upper left corner of the rectangle, `w` and `h` are width and height. `color` should be self-explanatory.

`[index]` is the array subscript operator. `enc[0]` is the first element of the array with the name `enc` defined earlier. `enc[1]` is the second element, etc. You probably want to display one bar for each encoder, so the first bar displays the value of `enc[0]`, the second, more to the right of `enc[1]`, the third `enc[2]`, the fourth `enc[3]`, you get the idea.

As written, the four rightmost bars on the display all show the position of the first encoder. You should add more elements to the `enc` array and then update the indices of the last four bars so they display the right value.

**Reply by laktosterror** [reply]
Oooh, now it make sense! Thank you for the explanation. I think i will be able to figure it out once the oleds arrive and i can experiment/test my way forward. I have seen you mention the "improved ssd1306" -library, is that library different from the "normal" ssd1306 library released by adafruit?

Thanks again for your time Pieter, let me get back to you with the results once the spi oleds arrive.

**Reply by tttapa** [reply]
> I have seen you mention the "improved ssd1306" -library, is that library different from the "normal" ssd1306 library released by adafruit?

I think this was a reference to my own fork of the Adafruit library: it supports using static buffers that can be shared between any number of displays. You won't really benefit from it unless you're using multiple OLED displays.

**Reply by laktosterror** [reply]
Ah I see. 

Got results! I tested my way forward and succeeded at last. 

If I'm using a teensy-pin for oled reset, does it have to have PWM? I'm trying pin 12 on a teensy LC (only pin free on my pcb) but no luck.

Thanks again for all your help, it means alot!! 

![PXL_20210830_223425514~2.jpg](https://user-images.githubusercontent.com/55037659/131415050-1ed98a57-0291-4bae-a403-f1aa54b6d965.jpg)

**Reply by tttapa** [reply]
> If I'm using a teensy-pin for oled reset, does it have to have PWM? I'm trying pin 12 on a teensy LC (only pin free on my pcb) but no luck.

No it doesn't have to be a PWM pin, any digital pin is fine ... except pin 12, because that is an SPI pin (MISO), so you cannot use it while the SPI interface is in use, even if you're only using it for output.

You can use an RC circuit instead of a GPIO pin to reset the display as explained in some of the OLED examples, but you might have to power cycle the board after uploading a new sketch for it to work.
